examples/bq_file_load_benchmark/benchmark_tools/file_generator.py

```diff
@@ -0,0 +1,662 @@
+            compression(str): String representing the compression format that
+                the generated file should have. Options are 'none' if no
+                compression is to be used, 'snappy', or 'deflate'.
+            extension(str): String to be used as the extension for the avro
+                file. Options are 'avro' if no compression is to be used,
+                'snappy', or 'deflate'.
+        """
+        pipeline_args = ['--project', self.project_id,
+                         '--staging_location', self.dataflow_staging_location,
+                         '--temp_location', self.dataflow_temp_location,
+                         '--save_main_session',
+                         '--worker_machine_type', 'n1-highcpu-32',
+                         '--runner', 'DataflowRunner',
+                         '--setup_file', './setup.py']
+        options = pipeline_options.PipelineOptions(pipeline_args)
+        table_spec = beam_bigquery.TableReference(
+            projectId=self.project_id,
+            datasetId=self.primitive_staging_dataset_id,
+            tableId=staging_table_util.table_id,
+        )
+        codec = 'null' if compression == 'none' else compression
+        bq_schema = staging_table_util.table.schema
+        table_name = staging_table_util.table.table_id
+        avro_schema = avro_util.AvroUtil(
+            bq_schema=bq_schema,
+            schema_name=table_name
+        ).get_avro_translated_schema()
+        p = beam.Pipeline(options=options)
+        table = (p
+                 | 'ReadTable' >> beam.io.Read(
+                     beam.io.BigQuerySource(table_spec)))
+        (table | beam.io.WriteToAvro(
+            file_path_prefix=destination_prefix,
+            schema=avro_schema,
+            file_name_suffix='.' + extension,
+            use_fastavro=True,
+            codec=codec,
+            num_shards=1,
+            shard_name_template='',
+        ))
+        p.run().wait_until_finish()
+        logging.info('Created file: {0:s}'.format(blob_name))
+
+    def _compose_sharded_blobs(self, blob_name):
+        """Composes multiple files (or blobs) into one file.
+
+        Args:
+            blob_name(str): Both the prefix that the sharded files share, and the
+                name of the single file that the sharded files will be composed
+                into.
+                Ex: fileType=csv/compression=none/numColumns=10/columnTypes=100_STRING/numFiles=10000/tableSize=2147MB/file3876.csv # pylint: disable=line-too-long
+        """
+        sharded_blobs = self.bucket.list_blobs(
+            prefix=blob_name
+        )
+        sharded_blobs = [blob for blob in sharded_blobs]
+        i = 0
+        precomposed_blobs = []
+
+        # Compose in groups of 32 'precomposed files', since no more than 32
+        # files can be compsed at once
+        logging.info('Composing sharded files into groups of 32.')
+        while i < len(sharded_blobs):
```

Comment:
Issues w/ this method:
1) Assumes that there are no more than 1024 files (32 * 32).
2) `precomposed_blob` is a confusing name because they are each themselves composed blobs.
3) Magic number 32
4) can parallelize compose and delete requests instead of blocking.


I think you can address these issues like this:
1) while loop 
2) `last_composed`
3) define constant
4) use a `ThreadPool.map`

Might look something like this

```python
def compose_blob(indexed_groups):
    i, name  = indexed_groups
    return self.bucket.blob('{0:s}_grouping{1:d}'.format(
                name, i))

MAX_COMPOSABLE_BLOBS = 32

indexed_groups = [(i % MAX_COMPOSABLE_BLOBS, 
                   sharded_blobs[i:i + MAX_COMPOSABLE_BLOBS])
                  for i in range(0, len(shared_blobs), MAX_COMPOSABLE_BLOBS)]

with ThreadPool() as p:
    indexed_composed = enumerate(p.map(compose_objects, indexed_groups))

    while len(indexed_composed) > 1:
        last_composed = [blob for _, blob in indexed_composed]
        indexed_composed = enumerate(p.map(compose_objecs, indexed_compose))
        # Clean up the artifacts from the last compose.
        p.map(self.bucket.delete_blob, last_composed)
    
    # Clean up original files.
    p.map(self.bucket.delete_blob, sharded_blobs)
```

Reply:
This could also use a test :)

Reply:
I totally understand the concerns listed above, but I'm having a bit of trouble following the example solution. For example, what purpose does the `i % MAX_COMPOSABLE_BLOBS` serve in `indexed_groups`? It should always be zero, right, since `i` is in increments of MAX_COMPOSABLE_BLOBS? For example, as a test, I put 7 sample blobs named blob1.csv, blob2.csv, and so on in a sample bucket. I set MAX_COMPOSABLE_BLOBS to 2. That gave me the following for `indexed_groups`: 
`[(0, [<Blob: annarudy_test2, blob1.csv>, <Blob: annarudy_test2, blob2.csv>]), (0, [<Blob: annarudy_test2, blob3.csv>, <Blob: annarudy_test2, blob4.csv>]), (0, [<Blob: annarudy_test2, blob5.csv>, <Blob: annarudy_test2, blob6.csv>]), (0, [<Blob: annarudy_test2, blob7.csv>])]`

Reply:
I can see where you would want to enumerate like you do to get `index_composed` to get the blobs into groups of size MAX_COMPOSABLE_BLOBS. But again, I'm not sure what the `grouping0` does: 
`[(0, <Blob: annarudy_test2, [<Blob: annarudy_test2, blob1.csv>, <Blob: annarudy_test2, blob2.csv>]_grouping0>), (1, <Blob: annarudy_test2, [<Blob: annarudy_test2, blob3.csv>, <Blob: annarudy_test2, blob4.csv>]_grouping0>), (2, <Blob: annarudy_test2, [<Blob: annarudy_test2, blob5.csv>, <Blob: annarudy_test2, blob6.csv>]_grouping0>), (3, <Blob: annarudy_test2, [<Blob: annarudy_test2, blob7.csv>]_grouping0>)]`

Reply:
Ok scratch those questions! I think the solution I came up with is pretty close to what you were going for.